File: Arcade/Intro/16.clj
(defn are_similar [a b ab]
  "are_similar : [Int], [Int], [Bool] -> Bool.
   Check similar from ab."
  (let [ab_index (map-indexed vector ab)
        false_pos (filter #(false? (second %)) ab_index)]
    (and (= (get a (first (first false_pos))) (get b (first (second false_pos))))
         (= (get a (first (second false_pos))) (get b (first (first false_pos)))))))

(defn areSimilar [a b]
  "areSimilar : [Int], [Int] -> Bool.
   Find similar Arrays, arrays are similar iff it is hequal by swapping
   only one pair of it elements."
  (if (= a b) true
      (let [eq_lst (map = a b)
            diff_a (clojure.set/difference (set a) (set b))
            diff_b (clojure.set/difference (set b) (set a))]
        (cond
          (and (not (empty? diff_a)) (not (empty? diff_b))) false
          (> (get (frequencies eq_lst) false) 2) false
          :otherwise (are_similar a b eq_lst)))))

File: Arcade/Intro/16.cpp
int diff_pos(std::vector<int> a, std::vector<int> b, int i = 0) {
  // diff_pos : [Int], [Int], Int -> Int.
  // Return the first position that elements of a and b are different
  // from i.
  for(; i < a.size(); i++)
    if(a[i] != b[i])
      return i;
  return -1;
}

bool areSimilar(std::vector<int> a, std::vector<int> b) {
  // areSimilar : [Int], [Int] -> Bool.
  // Find similar Arrays, arrays are similar iff it is equal by swapping
  // only one pair of it elements.
  int pos1, pos2;
  pos1 = diff_pos(a, b);
  if(pos1 == -1) return true;
  pos2 = diff_pos(a, b, pos1 + 1);
  if(pos2 == -1) return false;
  if(a[pos1] != b[pos2] || a[pos2] != b[pos1]) return false;
  return diff_pos(a, b, pos2 + 1) == -1;
}
